### ISSUE TITLE
'Application not available' when starting a Che workspace

### ISSUE THREAD

**Issue by sunix** [eclipse-che/che#17542]
### Describe the bug
Starting a workspace it happens randomly about 3 times a day in che.openshift.io

![image](https://user-images.githubusercontent.com/650571/89028767-47cfa100-d32d-11ea-96af-1dc087657fed.png)

I guess there is a synch issue, che-theia may be available internally but not externally maybe?

**Comment by ibuziuk** [comment]
This is a known issue that is supposed to be addressed in the upstream https://github.com/eclipse/che/issues/17002

**Comment by skabashnyuk** [comment]
I have some doubts about https://github.com/eclipse/che/issues/17002 and the current issue. Theia is a special server that is tested before reported that workspace is Running.

**Comment by sunix** [comment]
I am wondering if there is a global latency between the time the server is available from the same cluster and the time when it is available externally. I have also noticed that issue when opening an app preview in Che. it is happening randomly but still quite frequently.

**Comment by sunix** [comment]
or is readiness probe missing?

**Comment by ibuziuk** [comment]
> I have some doubts about #17002 and the current issue. Theia is a special server that is tested before reported that workspace is Running.

@skabashnyuk this might be related to route flakiness in this case and we might need to enable retries (e.g. mark theia as ready only after 3 successful checks). As I recall the retry functionality should be already in place and available for configuration

**Comment by ibuziuk** [comment]
The issue should be fixed on Hosted Che end by increasing the success threshold - https://github.com/redhat-developer/rh-che/pull/2009

**Comment by sunix** [comment]
@ibuziuk would that make sense to have that value the default one in Che ?

**Comment by ibuziuk** [comment]
@sunix it depends how easy it to repro against upstream + some default infra

**Comment by ibuziuk** [comment]
TL;DR this issue is infra related and the optimal config may vary

**Comment by sunix** [comment]
I frequently have that issue on openshift 4.4

**Comment by sunix** [comment]
frenquently => maybe 1/10 workspace start, but it is very random obviously.
it is just that it doesn't look like very robust when you live demo Che and have that issue.

**Comment by sunix** [comment]
it just happened to me on CodeReady Workspaces
![image](https://user-images.githubusercontent.com/650571/100066269-2a22cc00-2e35-11eb-948f-4e5462d37498.png)

**Comment by ibuziuk** [comment]
yeah, so we can increase it in the success threshold in the upstream, but it would result in slower performance - so it is a tradeoff.
Also, note that there will be no such an issue when single-host become a default route exposure strategy

**Comment by ibuziuk** [comment]
This is a central outage issue and if can not affect it. Closing since there was no issue reported during the periodic test summary.

**Comment by sunix** [comment]
I understand but I don't think changing threshold is the fix. The dashboard javascript client should check if the route works or not and retry.

**Comment by Katka92** [comment]
The periodic tests caught that again today.
[Job](https://ci.centos.org/job/devtools-rh-che-periodic-prod-2aProd/3624/console) and [screen](http://artifacts.ci.centos.org/devtools/rhche/devtools-rh-che-periodic-prod-2aProd/3624/e2e_report/report/Java_Maven_test_Create_Java_Maven_workspace_Wait_for_workspace_readiness/screenshot-Wait_for_workspace_readiness.png)

**Comment by sunix** [comment]
@ibuziuk ok to reopen that one ?

**Comment by ibuziuk** [comment]
@sunix  the real solution is single-host, we just need to wait till it becomes the default option. You can reopen, but I doubt that anything apart from what has been already done will be improved

**Comment by Katka92** [comment]
This issue was caught again by our periodic tests. [Job](https://ci.centos.org/job/devtools-rh-che-periodic-prod-preview-2a/3962/console) and [screen](http://artifacts.ci.centos.org/devtools/rhche/devtools-rh-che-periodic-prod-preview-2a/3962/e2e_report/report/Java_Maven_test_Create_Java_Maven_workspace_Wait_for_workspace_readiness/screenshot-Wait_for_workspace_readiness.png)

**Comment by sunix** [comment]
Let's reopen it and close it when we have a proper fix (single host or anything else)

**Comment by Katka92** [comment]
I've encountered that issue again when testing CRW.

**Comment by sleshchenko** [comment]
I faced a bit different but similar issue https://github.com/eclipse/che/issues/19059
I assume Dashboard itself can test IDE URL before opening iframe.

**Comment by ibuziuk** [comment]
@sleshchenko I put the UD label

**Comment by ibuziuk** [comment]
> I've encountered that issue again when testing CRW.

@Katka92 @sleshchenko folks, isn't this problem multi-host specific?

**Comment by Katka92** [comment]
IDK, I'm not load testing with single-host config. Maybe it would be worth adding that test case to load tests, but with lower priority than multi-host. Now I'm testing config with multi-host (~100 workspace startups for now) and I haven't seen that yet, but it was seen for testing CRW 2.9 (based on Che 7.30.x). I can try with single-host if I have time and let you know.

**Comment by ibuziuk** [comment]
@skabashnyuk do you have any input on when we will have a single-host enabled by default?

**Comment by skabashnyuk** [comment]
>@skabashnyuk do you have any input on when we will have a single-host enabled by default?

I don't know

**Comment by sleshchenko** [comment]
I think now single host is the default suggested exposure strategy, so it should not happen and from dashboard point of view it would be overhead. Closing it.
